pkg/cloud/rgraph/algo/parallel_queue.go

```diff
@@ -0,0 +1,241 @@
+/*
+Copyright 2023 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package algo
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"k8s.io/klog/v2"
+)
+
+type config struct {
+	workerCount int
+	tracer      Tracer
+}
+
+type QueueOption func(*config)
+
+func WorkerCount(n int) QueueOption  { return func(c *config) { c.workerCount = n } }
+func UseTracer(t Tracer) QueueOption { return func(c *config) { c.tracer = t } }
+
+type Tracer func(RunInfo)
+
+// RunInfo records the details of a task.
+type RunInfo struct {
+	// ID for the execution.
+	ID string
+	// Queued is the timestamp of when the task was queued.
+	Queued time.Time
+	// Start is the timestamp of when the task was started as a goroutine.
+	Start time.Time
+	// End is when the task was finished.
+	End time.Time
+	// Err is the result of the task.
+	Err error
+}
+
+// NewParallelQueue returns a new queue instance.
+func NewParallelQueue[T fmt.Stringer](opts ...QueueOption) *ParallelQueue[T] {
+	cfg := config{
+		workerCount: 2,
+		tracer:      func(RunInfo) {},
+	}
+	for _, o := range opts {
+		o(&cfg)
+	}
+	return &ParallelQueue[T]{
+		c:    cfg,
+		in:   make(chan struct{}, 100),
+		done: make(chan RunInfo, 100),
+	}
+}
+
+// ParallelQueue executes parallel tasks on separate goroutines.
+type ParallelQueue[T fmt.Stringer] struct {
+	c config
+
+	lock    sync.Mutex
+	state   queueState
+	pending []queueElement[T]
+	in      chan struct{}
+	done    chan RunInfo
+	active  int
+}
+
+type queueElement[T fmt.Stringer] struct {
+	ri   RunInfo
+	item T
+}
+
+type queueState int
+
+const (
+	stateNotStarted queueState = iota
```

Comment:
nit, I find using the atomic https://pkg.go.dev/sync/atomic values or ints easier to read for these things (and less prone to deadlock errors), instead of using a mutex to protect the `q.state` you just update the atomic type

Reply:
Maybe -- but we are grabbing the lock anyway, so for me, it's simpler than reasoning about atomic + lock.

Reply:
fair enough